doc/index.adoc

```diff
@@ -0,0 +1,146 @@
+
+[horizontal]
+int vertexCount()::
+
+
+    Return the count of vertices in the graph
+
+
+void forEachVertex(IntConsumer)::
+
+    Calls consumer for each vertex id
+
+
+PrimitiveIntIterator vertexIterator()::
+
+    Returns an vertex id Iterator
+
+
+int degree(int, Direction)::
+
+    Return the degree for the given vertex / direction
+
+void forEachEdge(int, Direction, EdgeConsumer) & ::
+
+Iterator<EdgeCursor> edgeIterator(int, Direction)::
+
+    Iterates over each edge of the given vertex / direction
+
+void forEachEdge(int, Direction, WeightedEdgeConsumer) & ::
+
+Iterator<WeightedEdgeCursor> weightedEdgeIterator(int, Direction)::
+
+    Iterates over each edge of the given vertex / direction.
+
+int toVertexId(long)::
+
+     Map node id to vertex id
+
+long toNodeId(int)::
+
+      Map vertex id back to it's original node id
+
+
+[NOTE]
+For better distinction between the Neo4j API and the Algorithm API we use the Terms Vertex
+and Edge rather then Node and Relation.
+
+The interface specifies Methods for iterating over all vertices of the
+graph as well as iterating over edges of a given vertex in the form of *forEach(..)*-
+and *iterator()*-Methods. The Graph knows the *nodeCount* and *degree* of each node and can *map*
+nodeId to vertexId and vice versa. An Iterator is implemented for (single-)weighted and unweighted edges.
+
+
+[ditaa]
+----
+
+                                  +--------------+
+                                  |    Graph     |
+                                  +--------------+
+                                  | - degree( )  |
+                                  | - forEach( ) |
+                                  | - iterator( )|
+                                  | - idMap( )   |
+                                  +--------------+
+                                    ^    ^     ^
+                        implements  |    |     | implements
+                      +-------------+    |     +----------+
+                      |                  |                |
+                      |                  |implements      |
+                 +----+-------+    +-----+------+     +---+-------+
+                 | HeavyGraph |    | LightGraph |     | GraphView |
+                 +------------+    +------------+     +-----------+
+
+----
+
+[NOTE]
+Currently we have 3 different implementations aiming for different goals like performance, memory
+consumption and accuracy.
+
+HeavyGraph::
+
+This implementations utilizes an int-matrix for storing connections between vertices.
+It has a high memory consumption but performs basic calculations on the graph
+around 3 times faster then the memory efficient implementation. Furthermore the
+number of edges per node is limited only to the maximum array size of the VM and
+connections can be added in arbitrary order.
+
+LightGraph::
+
+This implementation takes 3 times less heap due to a more intelligent
+memory layout. The drawback is the higher evaluation time.
```

Comment:
can we figure out why the light-graph is slower, from what I saw it shouldn't be.